spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,345 @@
+2.. The child process is sent a `Discover(Module|Resource)ReaderRequest` message.
+3.. A `DiscoverReadersResponse` message is received (within some timeout).
+7. The appropriate `Read*Request` message is sent to the child process, the corresponding `Read*Response` message is awaited.
+8. Evaluation continues.
+9. Evaluation completes.
+10. The `ExternalProcess` is closed, the child process is sent SIGTERM.
+11. If the child process has not terminated after some timeout (3 seconds?), it is forcefully stopped via SIGKILL.
+
+=== Java API
+
+New APIs:
+
+* `ResourceReaders.External` - a `ResourceReader` implementation identical to `ClientResourceReader`.
+* `ModuleKeyFactories.External` - a `ModuleKeyFactory` implementation similar to `ClientModuleKeyFactory`.
+* `ModuleKeys.External` - a `ModuleKey` implementation identical to `ClientModuleKey`.
+* `ExternalProcess` - manages the lifecycle of child processes.
+    ** Explicit `close` methods to manage child process lifecycle.
+    ** The `ExternalProcess` spawns the subprocess on first access, which sets up the `MessageTransport`, sends the appropriate `Discover*ReaderRequest` message, and awaits the corresponding `Discover*ReaderResponse` response.
+
+This proposal requires that the message passing API functionality move out of pkl-server and into pkl-core.
+The code added to pkl-core will include the new APIs and the core messaging code currently part of pkl-server (`pkl-server/src.main/kotlin/org.pkl.server/Message*.kt`).
+
+=== Message Passing API
+
+Four new message types will be added:
+
+[source,pkl]
+----
+/// Code: 0x100
+/// Type: Server Request
+class DiscoverModuleReaderRequest {
+  /// A number identifying this request
+  requestId: Int
+
+  /// The scheme of the resource to discover the spec for
+  scheme: String
+}
+
+/// Code: 0x101
+/// Type: Server Request
+class DiscoverResourceReaderRequest {
+  /// A number identifying this request
+  requestId: Int
+
+  /// The scheme of the resource to discover the spec for
+  scheme: String
+}
+
+/// Code: 0x102
+/// Type: Client Response
+class DiscoverModuleReaderResponse {
+  /// A number identifying this request
+  requestId: Int
+
+  /// Client-side module reader spec.
+  ///
+  /// Null when the external process does not implement the requested scheme.
+  /// [ClientModuleReader] is defined at https://pkl-lang.org/main/current/bindings-specification/message-passing-api.html#create-evaluator-request
+  spec: ClientModuleReader?
+}
+
+/// Code: 0x102
+/// Type: Client Response
+class DiscoverResourceReaderResponse {
+  /// A number identifying this request
+  requestId: Int
+
+  /// Client-side resource reader spec.
+  ///
+  /// Null when the external process does not implement the requested scheme.
+  /// [ClientResourceReader] is defined at https://pkl-lang.org/main/current/bindings-specification/message-passing-api.html#create-evaluator-request
+  spec: ClientResourceReader?
+}
+----
```

Comment:
Since the scheme is known, it's not so much of a discovery anymore. How about let's call this `InitializeModuleReader` and `InitializeResourceReader`?

Reply:
Updated